File: progettoP2/src/progettoP2/StatoPDA.java
package progettoP2;

import java.util.Vector;

public class StatoPDA extends Stato{
	private Vector<TransizionePDA> trans;
	public StatoPDA(){
		super(null, false);
		trans = new Vector<TransizionePDA>();
	}
	
	public StatoPDA(final String nome, final Boolean finale){
		super(nome,finale);
		trans = new Vector<TransizionePDA>();
	}
	
	public TransizionePDA getTrans(StatoPDA stato, final char input ,final char head,final String newHead){
		int j = 0;
	    while(j<trans.size() && (trans.elementAt(j).getDest() != stato || trans.elementAt(j).getInput() != input ||
	    		 trans.elementAt(j).getHead() != head || trans.elementAt(j).getNewHead().charAt(0) != newHead.charAt(0) ||
                 trans.elementAt(j).getNewHead().charAt(1) != newHead.charAt(1))){
	         j++;
	    }
	    if(j == trans.size()){
	        return null;
	    }
	    else{
	        return trans.get(j);
	    }
	}
	
	public void add(StatoPDA stato, final char input, final char head, final String newHead){
        if(getTrans(stato,input,head,newHead) == null)
        {
            trans.addElement(new TransizionePDA(stato,input,head,newHead));
        }
	}
	
	public void remove(StatoPDA stato,final char input, final char head, final String newHead){
		 Boolean eliminato = false;
	    for(int j = 0;  j < trans.size() && !eliminato; j++)
	    {
	        if(trans.get(j).getDest() == stato && 
	        		trans.get(j).getInput() == input 
	        		&& trans.get(j).getHead() == head 
	        		&& trans.get(j).getNewHead() == newHead)
	        {
	            trans.remove(j);
	            eliminato = true;
	        }
	    }
	}
	
	public int nTrans(){
		 return trans.size();
	}
	
	@Override
	public Transizione operatorBarraBarra(int n){
		 return trans.elementAt(n);
	}
}

File: ViewAutomata/statopda.cpp
#include "statopda.h"

StatoPDA::StatoPDA(const std::string & nome, const bool& finale):Stato(nome,finale)
{
    trans = new std::vector<TransizionePDA*>();
}

TransizionePDA *StatoPDA::getTrans(StatoPDA * stato, const char & input, const char & head, const std::string& newHead)
{
    std::vector<TransizionePDA*>::iterator i;
    for(i = trans->begin(); i != trans->end() && ((*i)->getDest() != stato || (*i)->getInput() != input ||
                             (*i)->getHead() != head || (*i)->getNewHead() != newHead); i++);
    if(i == trans->end())
        return nullptr;
    else
        return (*i);
}

void StatoPDA::add(StatoPDA * stato, const char& input, const char& head, const std::string& newHead)
{
    //if(newHead.size() <= 1 || (newHead.size() == 2 && head != '\0' && newHead[0] != '\0')) //accetto: s/s, s/e, e/s, e/e, s/ss
        if(getTrans(stato,input,head,newHead) == nullptr)
        {
            trans->push_back(new TransizionePDA(stato,input,head,newHead));
        }
}

void StatoPDA::remove(StatoPDA * stato, const char & input, const char & head, const std::string & newHead)
{
    bool eliminato = false;
    for(std::vector<TransizionePDA*>::iterator j = trans->begin();  j != trans->end() && !eliminato; j++)
    {
        if((*j)->getDest() == stato && (*j)->getInput() == input && (*j)->getHead() == head && (*j)->getNewHead() == newHead)
        {
            trans->erase(j);
            eliminato = true;
        }
    }
}

int StatoPDA::nTrans()
{
    return trans->size();
}

Transizione *StatoPDA::operator [](int i)
{
    return trans->at(i);
}
